packages/ens-utils/src/registrar.ts

```diff
@@ -0,0 +1,217 @@
+
+/**
+ * A single distinct charge that must be paid to a `Registrar` to register or
+ * renew a domain.
+ */
+export interface AbstractRegistrarCharge {
+
+  /**
+   * The type of `RegistrarCharge` that is being applied by the `Registrar` to
+   * rent the domain.
+   */
+  type: RegistrarChargeType;
+
+  /**
+   * The price of the `RegistrarCharge` that must be paid to the `Registrar` to
+   * rent the domain.
+   */
+  price: Price;
+
+  /**
+   * The reason why the `RegistrarCharge` is being applied by the `Registrar`
+   * to rent the domain.
+   */
+  reason?: string;
+
+  /**
+   * The period of time that the `RegistrarCharge` is valid for.
+   * 
+   * If `null`, the `RegistrarCharge` is not guaranteed to be time-limited.
+   */
+  validity?: TimePeriod;
+}
+
+export interface RegistrarBaseFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.BaseFee;
+  price: Price;
+  reason?: undefined;
+  validity?: undefined;
+}
+
+export interface RegistrarSpecialNameFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.SpecialNameFee;
+  price: Price;
+  reason: string;
+  validity?: undefined;
+}
+
+export interface RegistrarTemporaryFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.TemporaryFee;
+  price: Price;
+  reason: string;
+  validity: TimePeriod;
+}
+
+export interface RegistrarServiceFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.ServiceFee;
+  price: Price;
+  reason?: string;
+  validity?: undefined;
+}
+
+export type RegistrarCharge = RegistrarBaseFee | RegistrarSpecialNameFee | RegistrarTemporaryFee | RegistrarServiceFee;
+
+/**
+ * A price quote from a `Registrar` to register or renew a domain for a given
+ * period of time.
+ */
+export interface AbstractRegistrarPriceQuote {
+
+  /**
+   * The action associated with this `RegistrarPriceQuote`.
+   */
+  action: RegistrarAction;
+
+  /**
+   * The `TimePeriod` that this `RegistrarPriceQuote` is for.
+   */
+  rentalPeriod: TimePeriod;
+
+  /**
+   * The set of distinct `RegistrarCharge` values that must be paid to the
+   * `Registrar` to rent the domain for the period of `rentalPeriod`.
+   * 
+   * May be empty if the domain is free to rent for `rentalPeriod`.
+   */
+  charges: RegistrarCharge[];
+
+  /**
+   * The total price to rent the domain for the period of `rentalPeriod`.
+   * 
+   * This is the sum of `charges` in the unit of currency that should be paid
+   * to the registrar.
+   */
+  totalPrice: Price;
+}
+
+export interface RegistrationPriceQuote extends AbstractRegistrarPriceQuote {
+  action: typeof RegistrarAction.Register;
+}
+
+export interface RenewalPriceQuote extends AbstractRegistrarPriceQuote {
+  action: typeof RegistrarAction.Renew;
+}
+
+export interface Registrar {
+  getManagedSubname(name: ENSName): ENSName | null;
+
+  // NOTE: Throws RegistrarUnsupportedNameError if name is not supported by this registrar
```

Comment:
I am not sure I would leave this note here since this is already defined inside `getValidatedSubname` function

Reply:
This idea needs to be documented as part of the interface.